magento/magento2
# Implement "SELECT...FOR UPDATE" For Some Sales Transactions

Implementation of row level locking using "SELECT...FOR UPDATE" on the Order Repository and the `ShipOrder`, `RefundOrder`, and `InvoiceOrder` classes.


### Description (*)


Magento’s ORM does not support row-level out of the box. This allows multiple processes to attempt to update the same record at the same time, allowing partial data loss. When multiple updates for the same order are sent to Magento concurrently. It’s expected that concurrent updates for the same record would be isolated however, Magento allows these updates to overwrite each other (example below).
![image](https://user-images.githubusercontent.com/17521434/91856923-2ff18280-ec35-11ea-97c3-a1742b66f406.png)
In this screenshot you can see that Magento updates the `qty_invoiced` field in the `sales_order_item` table to reflect “Invoiced 8” when 10 invoices were sent to Magento and 10 invoices were created. This is a much smaller example, in other cases we’ve seen API calls be completely lost, where 10 invoice calls are sent to Magento and are successful, but we only see e.g 6 invoices created.
This issue was first reported in August of 2017 (https://github.com/magento/magento2/issues/10396), we can see that Adobe created an internal ticket `MAGETWO-82162`. However, it appears both tickets were closed without a solution in December of 2019. GitHub user @sshymko addressed the root cause of this issue stating, “Insufficient isolation between concurrent order update transactions is the root cause of the issue” and “Magento ORM does not protect from data integrity issues on concurrent requests to modify the same entity. For instance, repositories do not support loading with an intent to modify the entity. Basically, repositories have only a get() method, but nothing like getForUpdate() that would lock down the record via SELECT...FOR UPDATE for exclusive modification instead of a plain SELECT.” It’s unclear why his solution was never pursued by Adobe or the Magento open-source community.
I was able to solve this issue with the solution presented on GitHub: Implementing row-level locks into Magento’s ORM using ”SELECT…FOR UPDATE”

### Fixed Issues
1. Fixes magento/magento2#10396

### Manual testing scenarios (*)
1. Create Order With Multiple Lines E.G 15
2. Concurrently Ship and Invoice Order Lines 
3. See the related Order Detail Page, `qty_invoiced`, `qty_shipped` will be incorrect.

### Questions or comments
The following document has all of my test details/results and all information on the implementation and core issues behind it.
[Magento 2 Concurrent Processing Issues [FINAL].docx](https://github.com/magento/magento2/files/5156443/Magento.2.Concurrent.Processing.Issues.FINAL.docx)


### Contribution checklist (*)
